crates/cext/src/circuit.rs

```diff
@@ -0,0 +1,264 @@
+    };
+
+    let circuit = CircuitData::new(qubits, clbits, None, 0, (0.).into()).unwrap();
+    Box::into_raw(Box::new(circuit))
+}
+
+/// @ingroup QkCircuit
+/// Get the number of qubits the circuit contains.
+///
+/// @param circuit A pointer to the circuit.
+///
+/// @return The number of qubits the circuit is defined on.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     uint32_t num_qubits = qk_circuit_num_qubits(qc);  // num_qubits==100
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_num_qubits(circuit: *const CircuitData) -> u32 {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+
+    circuit.num_qubits() as u32
+}
+
+/// @ingroup QkCircuit
+/// Get the number of clbits the circuit contains.
+///
+/// @param circuit A pointer to the circuit.
+///
+/// @return The number of qubits the circuit is defined on.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     uint32_t num_qubits = qk_circuit_num_clbits(qc);  // num_qubits==100
+///
+/// # Safety
+///
+/// Behavior is undefined ``circuit`` is not a valid, non-null pointer to a ``QkCircuit``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_num_clbits(circuit: *const CircuitData) -> u32 {
+    // SAFETY: Per documentation, the pointer is non-null and aligned.
+    let circuit = unsafe { const_ptr_as_ref(circuit) };
+
+    circuit.num_clbits() as u32
+}
+
+/// @ingroup QkCircuit
+/// Free the circuit.
+///
+/// @param circuit A pointer to the circuit to free.
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100, 100);
+///     qk_circuit_free(qc);
+///
+/// # Safety
+///
+/// Behavior is undefined if ``circt`` is not either null or a valid pointer to a
+/// [CircuitData].
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_circuit_free(circuit: *mut CircuitData) {
+    if !circuit.is_null() {
+        if !circuit.is_aligned() {
+            panic!("Attempted to free a non-aligned pointer.")
+        }
+
+        // SAFETY: We have verified the pointer is non-null and aligned, so it should be
+        // readable by Box.
+        unsafe {
+            let _ = Box::from_raw(circuit);
+        }
+    }
+}
+
+/// @ingroup QkCircuit
+/// Append a standard gate to the circuit
+///
+/// @param circuit A pointer to the circuit to add the gate to
+/// @param gate The StandardGate to add to the circuit
+/// @param qubits The pointer to the array of ``uint32_t`` qubit indices to add the gate on
+/// @param params The pointer to the array of ``double`` values to use for the gate parameters
+///
+/// # Example
+///
+///     QkCircuit *qc = qk_circuit_new(100);
+///     qk_circuit_append_standard_gate(qc, HGate, *[0], *[]);
+///
+///
+/// # Safety
```

Comment:
Not actually sure the spacing here needs to be exact, feel free to ignore 
```suggestion
/// # Safety
///
```